pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -315,23 +440,166 @@ cdef inline timedelta_from_spec(object number, object frac, object unit):
+            return NotImplemented
+
+        try:
+            other = Timedelta(other)
+        except ValueError:
+            # failed to parse as timedelta
+            return NotImplemented
+
+        return Timedelta(op(self.value, other.value), unit='ns')
+
+    f.__name__ = name
+    return f
+
+
 # ----------------------------------------------------------------------
 # Timedelta Construction
 
+cdef _to_py_int_float(v):
+    # Note: This used to be defined inside _timedelta_value_kwargs
+    # (and Timedelta.__new__ before that), but cython
+    # will not allow `cdef` functions to be defined dynamically.
+    if is_integer_object(v):
+        return int(v)
+    elif is_float_object(v):
+        return float(v)
+    raise TypeError("Invalid type {0}. Must be int or "
+                    "float.".format(type(v)))
+
+
+cdef _timedelta_value_kwargs(dict kwargs):
+    # Helper for Timedelta.__new__
+    if not len(kwargs):
+        raise ValueError("cannot construct a Timedelta without a "
+                         "value/unit or descriptive keywords "
+                         "(days,seconds....)")
+
+    kwargs = {key: _to_py_int_float(kwargs[key]) for key in kwargs}
+
+    nano = kwargs.pop('nanoseconds', 0)
+    try:
+        value = convert_to_timedelta64(timedelta(**kwargs), 'ns') + nano
+    except TypeError as e:
+        raise ValueError("cannot construct a Timedelta from the "
+                         "passed arguments, allowed keywords are "
+                         "[weeks, days, hours, minutes, seconds, "
+                         "milliseconds, microseconds, nanoseconds]")
+    return value
+
+
 # Similar to Timestamp/datetime, this is a construction requirement for
 # timedeltas that we need to do object instantiation in python. This will
 # serve as a C extension type that shadows the Python class, where we do any
 # heavy lifting.
 cdef class _Timedelta(timedelta):
-    # cdef readonly:
-    #     int64_t value     # nanoseconds
-    #     object freq       # frequency reference
-    #     bint is_populated # are my components populated
-    #     int64_t _sign, _d, _h, _m, _s, _ms, _us, _ns
+    cdef readonly:
+        int64_t value      # nanoseconds
+        object freq        # frequency reference
+        bint is_populated  # are my components populated
+        int64_t _sign, _d, _h, _m, _s, _ms, _us, _ns
 
     # higher than np.ndarray and np.matrix
     __array_priority__ = 100
 
+    def __hash__(_Timedelta self):
+        if self._has_ns():
+            return hash(self.value)
+        else:
+            return timedelta.__hash__(self)
+
+    def __richcmp__(_Timedelta self, object other, int op):
+        cdef:
+            _Timedelta ots
+            int ndim
+
+        if isinstance(other, _Timedelta):
+            ots = other
+        elif PyDelta_Check(other):
+            ots = Timedelta(other)
+        else:
+            ndim = getattr(other, _NDIM_STRING, -1)
```

Comment:
is this the only use?

Reply:
of `_NDIM_STRING`?  Here, yes.  I think its used in Timestamp too.  Not totally sure what the benefit is.

Reply:
yeah this seems kind of silly

Reply:
fix this too